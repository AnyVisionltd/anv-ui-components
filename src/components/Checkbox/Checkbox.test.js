import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { getByRole } from '@testing-library/dom'
import { Checkbox } from './index'

describe('<Checkbox />', () => {
	describe('input state behaviour', () => {
		describe('checked state', () => {
			it('should be checked when set to be', () => {
				const { container } = render(<Checkbox checked />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.checked).toBe(true)
			})

			it('shouldn\'t be checked when not set to be', () => {
				const { container } = render(<Checkbox />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.checked).toBe(false)
			})
		})

		describe('indeterminate view state', () => {
			it('should be indeterminate when set to be', () => {
				const { container } = render(<Checkbox indeterminate />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.indeterminate).toBe(true)
			})

			it('shouldn\'t be indeterminate when not set to be', () => {
				const { container } = render(<Checkbox />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.indeterminate).toBe(false)
			})
		})

		describe('disabled state', () => {
			it('should be disabled when set to be', () => {
				const { container } = render(<Checkbox disabled />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.disabled).toBe(true)
			})

			it('shouldn\'t be disabled when not set to be', () => {
				const { container } = render(<Checkbox />)
				const checkbox = getByRole(container, 'checkbox')
				expect(checkbox.disabled).toBe(false)
			})
		})
	})

	describe('check click on checkbox functionality', () => {
		describe('normal checkbox', () => {
			describe('check input state change', () => {
				const { container } = render(<Checkbox />)
				const checkbox = getByRole(container, 'checkbox')

				it('should be changed from false to true after clicking on the checkbox', () => {
					expect(checkbox.checked).toBe(false)
					fireEvent.click(checkbox)
					expect(checkbox.checked).toBe(true)
				})

				it('should be changed from true to false after clicking on the checkbox', () => {
					expect(checkbox.checked).toBe(true)
					fireEvent.click(checkbox)
					expect(checkbox.checked).toBe(false)
				})
			})

			it('should trigger a change handler', async () => {
				const onChange = jest.fn()
				const { container } = render(<Checkbox onChange={ onChange } />)
				const checkbox = getByRole(container, 'checkbox')

				await userEvent.click(checkbox)
				expect(onChange).toBeCalled()
			})
		})

		describe('indeterminate checkbox', () => {
			describe('check input state change', () => {
				const { container } = render(<Checkbox indeterminate />)
				const checkbox = getByRole(container, 'checkbox')

				it('should be changed to checked after clicking on the checkbox', () => {
					expect(checkbox.indeterminate).toBe(true)
					expect(checkbox.checked).toBe(false)
					fireEvent.click(checkbox)
					expect(checkbox.checked).toBe(true)
				})

				it('should be changed to not checked after clicking on the checkbox again', () => {
					expect(checkbox.checked).toBe(true)
					fireEvent.click(checkbox)
					expect(checkbox.checked).toBe(false)
				})
			})

			it('should trigger a change handler', async () => {
				const onChange = jest.fn()
				const { container } = render(<Checkbox onChange={ onChange } />)
				const checkbox = getByRole(container, 'checkbox')

				await userEvent.click(checkbox)
				expect(onChange).toBeCalled()
			})
		})

		describe('disabled checkbox', () => {
			it('unchecked input should stay unchecked, change event should not be triggered', async () => {
				const onChange = jest.fn()
				const { container } = render(<Checkbox disabled onChange={ onChange } />)
				const checkbox = getByRole(container, 'checkbox')

				expect(checkbox.checked).toBe(false)
				await userEvent.click(checkbox)
				expect(checkbox.checked).toBe(false)
				expect(onChange).not.toBeCalled()
			})

			it('checked input should stay unchecked, change event should not be triggered', async () => {
				const onChange = jest.fn()
				const { container } = render(<Checkbox checked disabled onChange={ onChange } />)
				const checkbox = getByRole(container, 'checkbox')

				expect(checkbox.checked).toBe(true)
				await userEvent.click(checkbox)
				expect(checkbox.checked).toBe(true)
				expect(onChange).not.toBeCalled()
			})
		})
	})
})
