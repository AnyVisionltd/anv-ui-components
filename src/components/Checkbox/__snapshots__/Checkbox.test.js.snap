// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Button Default 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <button
      className="button large fill"
      disabled={false}
      onClick={[Function]}
    >
      Text
    </button>
  </div>
</div>
`;

exports[`Storyshots Button Disable 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill"
        disabled={false}
        onClick={[Function]}
      >
        Enable
      </button>
      <button
        className="button large fill"
        disabled={true}
        onClick={[Function]}
      >
        Disabled
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Button Play Ground 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <button
      className="button large fill"
      disabled={false}
      onClick={[Function]}
    >
      Button Text
    </button>
  </div>
</div>
`;

exports[`Storyshots Button Sizes 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill"
        disabled={false}
        onClick={[Function]}
      >
        Large
      </button>
      <button
        className="button small fill"
        disabled={false}
        onClick={[Function]}
      >
        small
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Button Variants 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill"
        disabled={false}
        onClick={[Function]}
      >
        fill
      </button>
      <button
        className="button large outline"
        disabled={false}
        onClick={[Function]}
      >
        outline
      </button>
      <button
        className="button large ghost"
        disabled={false}
        onClick={[Function]}
      >
        ghost
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Button With Class Name 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill successBackgroundColor"
        disabled={false}
        onClick={[Function]}
      >
        Text
      </button>
      <button
        className="button large outline alertColor"
        disabled={false}
        onClick={[Function]}
      >
        Text
      </button>
      <button
        className="button large ghost errorColor"
        disabled={false}
        onClick={[Function]}
      >
        Text
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots Button With Icon 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <button
      className="button large fill"
      disabled={false}
      onClick={[Function]}
    >
      <span
        className="leadingIcon"
      >
        <svg />
      </span>
      leading
    </button>
  </div>
</div>
`;

exports[`Storyshots Checkbox Colors 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <label
      className="checkbox primary checkbox-primary checked microMargin"
    >
      <input
        checked={true}
        className="parent"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      >
        <svg />
      </span>
    </label>
  </div>
</div>
`;

exports[`Storyshots Checkbox Default 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <label
      className="checkbox primary checkbox-primary"
      htmlFor="checkbox-example"
    >
      <input
        checked={false}
        className="parent"
        disabled={false}
        id="checkbox-example"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      />
    </label>
    <label
      className="checkboxLabel"
      htmlFor="checkbox-example"
    >
      Check me out!
    </label>
  </div>
</div>
`;

exports[`Storyshots Checkbox Playground 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <label
      className="checkbox primary checkbox-primary"
      htmlFor="checkbox-playground"
    >
      <input
        checked={false}
        className="parent"
        disabled={false}
        id="checkbox-playground"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      />
    </label>
    <label
      className="checkboxLabel"
      htmlFor="checkbox-playground"
    >
      Hit me!
    </label>
  </div>
</div>
`;

exports[`Storyshots Checkbox States 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <label
      className="checkbox primary checkbox-primary checked microMargin"
    >
      <input
        checked={true}
        className="parent"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      >
        <svg />
      </span>
    </label>
    <label
      className="checkbox primary checkbox-primary microMargin"
    >
      <input
        checked={false}
        className="parent"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      />
    </label>
    <label
      className="checkbox primary checkbox-primary indeterminate microMargin"
    >
      <input
        checked={false}
        className="parent"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      >
        <svg />
      </span>
    </label>
    <label
      className="checkbox primary checkbox-primary isDisabled microMargin"
    >
      <input
        checked={false}
        className="parent"
        disabled={true}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="iconContainer"
      />
    </label>
  </div>
</div>
`;

exports[`Storyshots IconButton Default 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <button
      className="button large fill iconButton small fill"
      disabled={false}
      onClick={[Function]}
    >
      <svg />
    </button>
  </div>
</div>
`;

exports[`Storyshots IconButton Disable 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill iconButton small fill"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large fill iconButton small fill"
        disabled={true}
        onClick={[Function]}
      >
        <svg />
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots IconButton Play Ground 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <button
      className="button large fill iconButton small fill"
      disabled={false}
      onClick={[Function]}
    >
      <svg />
    </button>
  </div>
</div>
`;

exports[`Storyshots IconButton Sizes 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill iconButton small fill"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large fill iconButton large fill"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots IconButton Variants 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill iconButton small fill"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large outline iconButton small outline"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large ghost iconButton small ghost"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
    </div>
  </div>
</div>
`;

exports[`Storyshots IconButton With Class Name 1`] = `
<div
  className="story-wrapper"
>
  <div
    className="center"
  >
    <div
      className="marginFlexContainer"
    >
      <button
        className="button large fill iconButton small fill successBackgroundColor"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large outline iconButton small outline alertColor"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
      <button
        className="button large ghost iconButton small ghost errorColor"
        disabled={false}
        onClick={[Function]}
      >
        <svg />
      </button>
    </div>
  </div>
</div>
`;
